### ISSUE TITLE
improve python gradient descent

### ISSUE THREAD

**Issue by suntzu86** [Yelp/MOE#91]
Right now, gradient descent just runs its iterations and stops. It'd be nice to consider a few more items:
1) stop when tolerance is reached (so we don't waste time moving 1.0e-14 in circles); this can be a check on the gradient size OR the step-size (e.g., if gradient is small-ish AND `pre_mult*gama^-n` is small, we won't make anymore meaningful progress).
2) restarts worked well for me in C++; we could try this in Python too (after GD exists, kick off another GD run from the final coordinates unless convergence was reached). this lets you be more conservative with the learning rate control & the algorithm will still be robust

1, 2) are modifications to the GD code. This is already done in C++ so we can model after that.
